cuda/base/kernel_launch_solver.cuh

```diff
@@ -30,45 +30,49 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************<GINKGO LICENSE>*******************************/
 
-#include "core/solver/ir_kernels.hpp"
+#ifndef GKO_CUDA_BASE_KERNEL_LAUNCH_SOLVER_CUH_
+#define GKO_CUDA_BASE_KERNEL_LAUNCH_SOLVER_CUH_
 
 
-#include <ginkgo/core/base/math.hpp>
-
-
-#include "cuda/components/thread_ids.cuh"
+#include "common/base/kernel_launch_solver.hpp"
 
 
 namespace gko {
 namespace kernels {
 namespace cuda {
-/**
- * @brief The IR solver namespace.
- *
- * @ingroup ir
- */
-namespace ir {
-
-
-constexpr int default_block_size = 512;
 
 
-#include "common/solver/ir_kernels.hpp.inc"
+template <typename KernelFunction, typename... KernelArgs>
+__global__ __launch_bounds__(default_block_size) void generic_kernel_2d_solver(
+    size_type rows, size_type cols, size_type default_stride, KernelFunction fn,
+    KernelArgs... args)
+{
+    auto tidx = thread::get_thread_id_flat();
+    auto col = tidx % cols;
+    auto row = tidx / cols;
+    if (row >= rows) {
+        return;
+    }
+    fn(row, col,
+       device_unpack_solver_impl<KernelArgs>::unpack(args, default_stride)...);
+}
 
 
-void initialize(std::shared_ptr<const CudaExecutor> exec,
-                Array<stopping_status> *stop_status)
+template <typename KernelFunction, typename... KernelArgs>
+void run_kernel_solver(std::shared_ptr<const CudaExecutor> exec,
+                       KernelFunction fn, dim<2> size, size_type default_stride,
+                       KernelArgs &&... args)
 {
-    const dim3 block_size(default_block_size, 1, 1);
-    const dim3 grid_size(ceildiv(stop_status->get_num_elems(), block_size.x), 1,
-                         1);
-
-    initialize_kernel<<<grid_size, block_size, 0, 0>>>(
-        stop_status->get_num_elems(), stop_status->get_data());
+    gko::cuda::device_guard guard{exec->get_device_id()};
```

Comment:
This is mainly relevant if we later decide to add this capability also for users, otherwise they might execute on the wrong GPU if this is not executed in the context of an `Operation`.